Baikal/Kernels/CL/scene.cl

```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
         diffgeo->ng = -diffgeo->ng;
     }
 
-    /// Calculate tangent basis
-    /// From PBRT book
-    float du1 = uv0.x - uv2.x;
-    float du2 = uv1.x - uv2.x;
-    float dv1 = uv0.y - uv2.y;
-    float dv2 = uv1.y - uv2.y;
-    float3 dp1 = v0 - v2;
-    float3 dp2 = v1 - v2;
-    float det = du1 * dv2 - dv1 * du2;
-
-    if (0 && det != 0.f)
+    // Calculate position and normal derivatives of the surface with respect to texcoord
+    // From PBRT book
+
+    /*
+        NOTE:
+        In the previous version, dpdu and dpdv vectors in the
+        structure DifferentialGeometry treated as tangent
+        and bitangent vectors in the kernels.
+        Actually, there is a difference: dpdu/dpdv have to be
+        oriented along increasing of uv coordinates, may be not
+        orthogonal and their length may be not equal to one.
+        We use these values for computation of screen space uv
+        derivatives.
+        On the other hand, normal, tangent and bitangent vectors
+        should be orthonormal, because we can easily get inversion of
+        tangent to world space transformation matrix simply by
+        transposing it.
+        So now we distinguish dpdu/dpdv and tangent/bitangent vectors.
+    */
+
+    float2 duv02 = uv0 - uv2;
+    float2 duv12 = uv1 - uv2;
+    float  det = duv02.x * duv12.y - duv02.y * duv12.x;
+    bool   degenerate_uv = fabs(det) < 1e-08f;
```

Comment:
please, add constant for this value (FLT_EPSILON for example) and use it.

Reply:
Done